### PR TITLE
[flink] Preserve log-only source restore mode

### LINKED ISSUE #3354: [flink] Avoid initializing lake snapshot splits after log-only source restore
### Search before asking

- [x] I searched in the [issues](https://github.com/apache/fluss/issues) and found nothing similar.

### Fluss version

main (development)

### Please describe the bug 🐞

When a Flink source starts from a specified log position, such as timestamp/latest/earliest startup instead of full startup, `FlinkTableSource` correctly passes `null` as the `lakeSource` to `FlinkSourceEnumerator`. The source therefore initializes only Fluss log splits and does not initialize lake snapshot splits.

After the job is restored from a checkpoint/savepoint, the recreated source can pass a non-null `lakeSource` to `FlinkSourceEnumerator`. Because the enumerator checkpoint state currently stores `remainingHybridLakeFlussSplits` as `null` in the original log-only startup, the restored enumerator treats the state as if lake splits have not been generated yet. It then initializes lake snapshot splits during restore.

This changes the startup semantics across restore and can duplicate historical data: the original run starts from the specified Fluss log position, but the restored run additionally reads lake snapshot data.

Expected behavior: restoring a source that originally started in log-only mode should preserve that mode and must not initialize lake snapshot splits.

Actual behavior: restore can initialize lake snapshot splits if `lakeSource` becomes non-null, causing duplicated initialization/read of historical lake data.

### Solution

Use the existing `remainingHybridLakeFlussSplits` checkpoint field as a sentinel. For a source enumerator that starts with lake snapshot reading disabled, persist an empty list instead of `null`. On restore, `pendingHybridLakeFlussSplits != null` already means the enumerator should not list/generate lake splits again, so an empty list preserves the original log-only startup semantics without changing the enumerator state schema.

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by luoyuxia**
### Purpose

Linked issue: close #3354

Preserve the original Fluss-only (non-lake) startup semantics after a Flink source restore. When the source starts from a specified Fluss log position, it should not initialize lake snapshot splits after restore just because `lakeSource` is available again.

### Brief change log

- Persist an empty `remainingHybridLakeFlussSplits` list when the enumerator checkpoints with `lakeSource == null`.
- Allow empty pending hybrid split lists to deserialize without `lakeSource`, while still requiring `lakeSource` for non-empty hybrid split state.
- Document the three-state meaning of `pendingHybridLakeFlussSplits` so `null`, empty, and non-empty states are explicit.
- Add a restore regression test that checkpoints a Fluss-only enumerator, restores it with a non-null `lakeSource` and a registered lake snapshot, and verifies no `LakeSnapshotSplit` is generated.
- Add a serializer round-trip test for empty pending hybrid split state with `lakeSource == null`.

### Tests

- `./mvnw -pl fluss-flink/fluss-flink-common -Dtest=SourceEnumeratorStateSerializerTest,FlinkSourceEnumeratorTest#testRestoreFlussOnlySourceWithLakeSourceDoesNotGenerateLakeSplits test`
- `./mvnw -pl fluss-flink/fluss-flink-common spotless:apply`
- `git diff --check`

### API and Format

No public API, storage format, or checkpoint serializer version change. The fix only changes the value stored in the existing enumerator state field for Fluss-only source checkpoints and permits that empty-state sentinel to deserialize without a lake source.

### Documentation

No new feature or user-facing configuration change.

### Generative AI disclosure

- [x] Yes: OpenAI Codex was used to author this PR following the repository AGENTS.md guidance.

### CODE REVIEW

**Comment by luoyuxia** [comment]
@loserwang1024 Could you please help review

File: fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/source/enumerator/FlinkSourceEnumerator.java [hunk 1]
```diff
@@ -1207,11 +1220,16 @@ public void addReader(int subtaskId) {
 
     @Override
     public SourceEnumeratorState snapshotState(long checkpointId) {
+        List<SourceSplitBase> remainingHybridLakeFlussSplits =
+                // Preserve Fluss-only (non-lake) startup across restore. Otherwise a restored
+                // enumerator with a non-null lakeSource would treat null as "not initialized yet"
+                // and generate lake snapshot splits.
+                lakeSource == null ? Collections.emptyList() : pendingHybridLakeFlussSplits;
```
Comment (loserwang1024) [inline fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/source/enumerator/FlinkSourceEnumerator.java]: I understand what you mean, but the code seems hard to understand without sufficient context. I've also thought about this:

- If the enumerator is created by FlinkSource#createEnumerator, it indicates a stateless restart. Therefore, whether to generate lake splits depends on whether it's a LakeSource.

- If the enumerator is created by FlinkSource#restoreEnumerator, there's no need to generate lake splits again. This is because before the first checkpoint is taken, FlinkSourceEnumerator#start → FlinkSourceEnumerator#generateHybridLakeFlussSplits has already been executed. Thus, upon restoration, the lake splits do not need to be regenerated.

Therefore, even if the job was previously started from a specified timestamp, according to this logic, as long as a checkpoint has been taken, upon stateful restart it will not read the lake splits again.

Comment (luoyuxia) [reply fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/source/enumerator/FlinkSourceEnumerator.java]: I agree the original approach in snapshotState() is hard to understand without sufficient context.

I've reworked the fix to use the checkpointTriggeredBefore flag in generateHybridLakeFlussSplits() instead. While the restore-awareness logic still requires some thought, this approach keeps all the complexity contained within a single method rather than spreading it across snapshotState().

Additionally, I changed startInStreamModeForNonPartitionedTable to call generateHybridLakeFlussSplits() synchronously, consistent with the partitioned-table path in start(). This ensures lake split initialization always completes before any checkpoint can be triggered, which is a prerequisite for the checkpointTriggeredBefore guard to work correctly.

**Comment by luoyuxia** [comment]
@loserwang1024 Comments has been addressed.

File: fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/source/enumerator/FlinkSourceEnumerator.java [hunk 2]
```diff
@@ -869,13 +881,19 @@ private List<SourceSplitBase> getLogSplit(
...
+        // Restored from checkpoint with pending lake splits — return them directly
+        // without re-generating.
         if (pendingHybridLakeFlussSplits != null) {
             LOG.info("Still have pending lake fluss splits, shouldn't list splits again.");
             return pendingHybridLakeFlussSplits;
         }
+        // Restored from checkpoint but pending lake split is null(e.g. the source was
+        // originally started in Fluss-only mode without lake).  Do not generate lake
+        // splits for this restore; mark as initialized and return empty list.
+        if (checkpointTriggeredBefore) {
```
Comment (fresh-borzoni) [inline fluss-flink/fluss-flink-common/src/main/java/org/apache/fluss/flink/source/enumerator/FlinkSourceEnumerator.java]: I think the checkpointTriggeredBefore guard is solving the restore ambiguity in the wrong layer.

Fresh null means “not initialized yet”.
Restored null should be normalized to “nothing pending”.

That normalization belongs in FlinkSource.restoreEnumerator(), not in generateHybridLakeFlussSplits().
WDYT?

**Review by fresh-borzoni — COMMENTED** [review]
@luoyuxia Ty for the PR, I agree with the sync change and the serializer reorder, but I have comment about checkpointTriggeredBefore guard.

**Comment by luoyuxia** [comment]
cc @loserwang1024 @loserwang1024 Please help review again.

**Review by fresh-borzoni — APPROVED** [review]
@luoyuxia Ty, LGTM 👍

**Comment by luoyuxia** [comment]
@loserwang1024 I'm going to merge this. Feel free to left any comment if you still have
